### DISCUSSION TITLE
Thermostat support on Matter

### DISCUSSION THREAD

**Discussion by devtekve** [arendst/Tasmota#19175] [Ideas]
Hello everyone!, I've been playing a bit with Tasmota coming from ESPHome, I still have a lot to learn! 

I have seen the big contributions from @s-hadinger regarding matter support on Tasmota (quite impressive stuff!) and became curious, I do not see any thermostat device type supported on Tasmota yet... I also tried to search for other issues / mrs / branches / discussions regarding supporting the Thermostat on Tasmota for matter but couldn't really find any... 

So I have a couple questions if anybody can answer them :D 

1. Is thermostat support on Tasmota via Matter planned and I did a terrible job searching?
2. When looking at the source code for matter, I see it is located under `berry_matter`, is that the _only_ implementation done and planned? is there a reason why `Berry` was chosen in particular? (_Don't get me wrong, I am not criticizing the choice, I am eager to learn the rationale behind those decisions_) 
3. I see two folders, `Embedded` and `Solidyfy`, while `Embedded` seems the clearest I am puzzled a bit about the `Solidify`, it seems like bytecode but written as c code... Is this automatically generated or manually written? 
4. Where should I start trying to implement the thermostat for Matter? 

Thank you for you time reading!

**Comment by s-hadinger** [comment]
Hi,

Thermostat is not yet supported. It's mostly implementing by needs, so I don't do it unless somebody is interested.

Indeed Matter support is located in `berry_matter` as a lib, which avoids polluting the main directory of drivers. Berry allows to develop **much faster** compared to C++, and performance is mostly not an issue with Matter.

The code is indeed in `embedded`. Then the code is compiled into bytecode (automatically) and the bytecode is stored in Flash. This avoids consuming RAM which is a very previous resource on ESP32. To solidify, you just need to run `./solidify_all.be` in `berry_matter`. But before you need to have a local version of Berry: in `berry` directory, just do `make`. For windows users, compiling Berry can be challenging so a pre-compiled `berry.exe` is provided.

You might also be interested in reading https://tasmota.github.io/docs/Matter-Internals/ which is still incomplete but explains the Plug-in structure to implement new device types.

It's not super easy to add a plug-in right now, and documenting it is not high in my priorities...

**Comment by devtekve** [comment]
Thank you very much for replying back!

Currently I use ESPHome because it was the easiest to get working with a
custom integration (I have reversed engineered an old thermostat controller
that was in my home), and I tried to play a little with berry but the loop
time I needed had to be much short (500 micro seconds), hence my question
about the usage of berry because my experience with it was limited, so I’m
glad to hear that there’s a lot of potential on it!!

I will take a look at the docs provided and keep playing with it. Thank you
for your time !

**Reply by s-hadinger** [reply]
I looked at Matter Thersmostat, and there are many attributes. I think we can limit to the smallest subset.

My problem is that I haven't found any standard Thermostat feature in Tasmota. So I don't have a natural mapping for it. What I can do is create a "virtual thermostat" that gets the command, and then you add scripting or rules to make it whatever you want.

What kind of control do you have on your thermostat?

**Reply by devtekve** [reply]
Thank you for you help ! I highly appreciate it :) 

> I haven't found any standard Thermostat feature in Tasmota

I found the following for the Thermostat feature in Tasmota: https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_xdrv_driver/xdrv_39_thermostat.ino but that one is extremely long and feature packed, so it might be a big challenge. The virtual thermostat sounds like a good starting point. 

> What kind of control do you have on your thermostat?

My current implementation could be seen as a simple relay... But instead of a relay, I send pulses through a GPIO pin to request the AC to turn ON or OFF. My ESP device will control the temperature. Currenlty with ESPHome I read the temperature using the HomeAssistant API, so that will have to change for Tasmota, and MQTT could be a way... Though my end goal would be supporting natively matter instead of MQTT or Home Assistant's API 

## A lot of extra context that might help 

This is the code I have that works to turn on and off the AC within Tasmota (please note there's no temperature reading implemented, and just the "relay" behavior): 

### MasterzoneManager.h
```cpp
#pragma once
// Time for each bit in microseconds
#define bit_time 500 

// Configure the given pin as output and set it high
void configure_pin(uint32_t provided_pin);

// Send a signal to the AC unit
void send_signal_to_ac(int pin, int *signal, int len, int elSize, int rpt);

// Turn the AC on by sending the 'on' signal
void turn_on_ac();

// Turn the AC off by sending the 'off' signal
void turn_off_ac();
```

###  MasterzoneManager.cpp
```cpp
#include <Arduino.h>
#include "MasterzoneManager.h"

#define bit_time 500 // Time for each bit in microseconds

// Array to represent the signal for turning the AC on
int on_signal[] = {0,0,0,0,0,0,1,1,0,1,1,1,0,0,0,1,0,0,0,1,0,0,0,1,0,0,0,1,0,0,0,1,0,0,0,1,0,1,1,1,0,1,1,1,0,0,0,1,0,0,0,1,0,0,0,1,0,0,0,1,0,0,0,1,0,0,0,1,0,1,1,1};

// Array to represent the signal for turning the AC off
int off_signal[] = {0,0,0,0,0,0,1,1,0,1,1,1,0,0,0,1,0,0,0,1,0,0,0,1,0,1,1,1,0,0,0,1,0,0,0,1,0,0,0,1,0,0,0,1,0,1,1,1,0,0,0,1,0,0,0,1,0,0,0,1,0,1,1,1,0,0,0,1,0,0,0,1};

// The GPIO pin connected to the receiver, default to GPIO4
volatile uint32_t  pin = 4;           

// Configure the given pin as output and set it high
void configure_pin(uint32_t provided_pin)
{
  pin = provided_pin;
  pinMode(pin, OUTPUT);
  digitalWrite(pin, HIGH);
}

// Send a signal to the AC unit
void send_signal_to_ac(int pin, int *signal, int len, int elSize, int rpt)
{
    // Repeat the signal 'rpt' times
    for (int ml = 0; ml < rpt; ++ml)
    {
        // Loop over each element in the signal array
        for (int i = 0; i <= len / elSize; i++)
        {
            // Write the signal to the pin
            digitalWrite(pin, signal[i]);

            // Only delay if we are NOT at the end
            if (i < (len / elSize))
                delayMicroseconds(bit_time);
        }
    }
}

// Turn the AC on by sending the 'on' signal
void turn_on_ac()
{
    digitalWrite(pin, HIGH);
    send_signal_to_ac(pin, on_signal, sizeof(on_signal), sizeof(on_signal[0]), 4);
    digitalWrite(pin, HIGH);
}

// Turn the AC off by sending the 'off' signal
void turn_off_ac()
{
    digitalWrite(pin, HIGH);
    send_signal_to_ac(pin, off_signal, sizeof(off_signal), sizeof(on_signal[0]), 4);
    digitalWrite(pin, HIGH);
}
```

### xdrv_102_masterzone.ino
And this is my custom driver on Tasmota 
```cpp
#ifdef USE_MASTERZONE
#include "MasterzoneManager.h"

#warning **** USE_MASTERZONE Driver is included... ****

#define XDRV_102 102

#define MASTERZONE_RX_PIN  4 

bool initSuccess = false;
uint32_t inputSwitchNumber = 1;   // Input switch number

const char MyProjectCommands[] PROGMEM = "|"
  "Help|"
  "TurnAcOff|"
  "TurnAcOn";

void (* const MyProjectCommand[])(void) PROGMEM = {&CmdHelp, &CmdTurnAcOff, &CmdTurnAcOn};

void CmdHelp(void) {
  AddLog(LOG_LEVEL_INFO, "Available commands:");
  AddLog(LOG_LEVEL_INFO, "TurnAcOff");
  AddLog(LOG_LEVEL_INFO, "TurnAcOn");
  AddLog(LOG_LEVEL_INFO, "Help");
}

void CmdTurnAcOn(void) {
  AddLog(LOG_LEVEL_INFO, PSTR("CmdTurnAcOn"));
  turn_on_ac();
}

void CmdTurnAcOff(void) {
  AddLog(LOG_LEVEL_INFO, PSTR("CmdTurnAcOff"));
  turn_off_ac();
}

void MasterzoneManagerInit() {
  configure_pin(MASTERZONE_RX_PIN);
  initSuccess = true;
  AddLog(LOG_LEVEL_INFO, "Masterzone init successful");
}

void MasterzoneProcessing() {
  AddLog(LOG_LEVEL_DEBUG, PSTR("MasterzoneProcessing"));
  if(!initSuccess)
    return;

  // this switchState is not currently working as I expected, so far this is only triggering the ON or OFF when I issue the command via the console
  AddLog(LOG_LEVEL_DEBUG, PSTR("setting switchState for [%d]"), inputSwitchNumber);
  uint32_t switchState = SwitchGetState(inputSwitchNumber);

  AddLog(LOG_LEVEL_DEBUG, PSTR("switchState: [%d]"), switchState);
  if(switchState)
    CmdTurnAcOn();
  else
    CmdTurnAcOff();
}

bool Xdrv102(uint32_t function) {
  bool result = false;

  if (FUNC_INIT == function) {
    MasterzoneManagerInit();
  } else if (initSuccess) {
    switch (function) {
      case FUNC_EVERY_SECOND:
        MasterzoneProcessing();
        break;
      case FUNC_COMMAND:
        result = DecodeCommand(MyProjectCommands, MyProjectCommand);
        ResponseCmndDone();
        break;
    }
  }

  return result;
}

#endif  // USE_MASTERZONE
```

**Reply by s-hadinger** [reply]
I'm leaving for more than 2 weeks, so we'll see this later.

One quick comment on the code above:
```
int on_signal[] = {0,0...
```
is not recommended since it creates an array in RAM, which is a very previous resource.

To force in Flash, use instead:
```
const int on_signal[] = {0,0...
```

Also you should consider a more compact way of encoding this array of bits. At least reduce to `uint8_t` instead of wasting full 32 bits for a single bit.
